Feature/application validators

AlexBud369 opened the pull request:
### Authentication Validators
- **`RegisterCommandValidator` (POST /auth/register)**: Validates `Email`, `Password`, `UserName`. Ensures valid email (max 100 chars), complex password (min 6 chars, uppercase, number), and username (max 50 chars, alphanumeric, underscores). Prevents insecure registration data.
- **`LoginCommandValidator` (POST /auth/login)**: Validates `Email` (valid, max 100 chars) and `Password` (non-empty, min 6 chars). Ensures valid credentials for JWT generation.

### User Validators
- **`BlockUserCommandValidator` (POST /users/{id}/block)**: Validates `UserId`, `AdminId`. Ensures non-empty IDs and `UserId` ≠ `AdminId`. Secures user blocking for admins.
- **`ChangeRoleCommandValidator` (POST /users/{id}/role)**: Validates `UserId`, `AdminId`, `Role` ("User" or "Admin"). Ensures non-empty IDs and valid role. Secures role assignment.
- **`DeleteUserCommandValidator` (DELETE /users/{id})**: Validates `UserId`, `AdminId`. Ensures non-empty IDs and `UserId` ≠ `AdminId`. Secures user deletion for admins.
- **`GetUserByIdQueryValidator` (GET /users/{id})**: Validates `Id`. Ensures non-empty ID for user retrieval.
- **`GetAllUsersQueryValidator` (GET /users)**: Validates `PageNumber` (> 0), `PageSize` (1–100), `Email` (max 100 chars), `UserName` (max 50 chars), `IsBlocked`. Ensures valid pagination and filtering.

### Tag Validators
- **`CreateTagCommandValidator` (POST /tags)**: Validates `TagName` (non-empty, max 50 chars, alphanumeric, spaces, hyphens). Prevents invalid tag creation.
- **`GetTagByIdQueryValidator` (GET /tags/{id})**: Validates `Id`. Ensures non-empty ID for tag retrieval.
- **`GetAllTagsQueryValidator` (GET /tags)**: Validates `PageNumber` (> 0), `PageSize` (1–100), `TagName` (max 50 chars, alphanumeric, spaces, hyphens). Ensures valid pagination and filtering.

### Book Validators
- **`CreateBookCommandValidator` (POST /books)**: Validates `BookCreateDto` (`Title`, `Author` non-empty, max 200/100 chars; `Description` optional, max 1500 chars; `Tags` non-empty, max 50 chars). Ensures valid book creation.
- **`UpdateBookCommandValidator` (PUT /books/{id})**: Validates `BookUpdateDto`, `Id`, `UserId`. Ensures at least one field updated, non-empty IDs. Secures book updates.
- **`DeleteBookCommandValidator` (DELETE /books/{id})**: Validates `Id`, `UserId`. Ensures non-empty IDs for authorized book deletion.
- **`UploadBookCoverCommandValidator` (POST /books/{id}/cover)**: Validates `Id`, `UserId`, `CoverImageUrl` (valid URL, max 500 chars, or `IFormFile` in future). Secures cover uploads.
- **`GetBookByIdQueryValidator` (GET /books/{id})**: Validates `Id`. Ensures non-empty ID for book retrieval.
- **`GetAllBooksQueryValidator` (GET /books)**: Validates `PageNumber` (> 0), `PageSize` (1–100), `SearchTerm` (max 200 chars), `Tag` (max 50 chars), `SortBy` (valid fields). Ensures valid pagination, filtering, sorting.

AlexBud369 committed:
chore: add FluentValidation and MediatR packages
- added FluentValidation and FluentValidation.DependencyInjectionExtensions to Application project for validator implementation
- added MediatR and MediatR.Extensions.Microsoft.DependencyInjection for CQRS and ValidationBehavior
- added AutoMapper.Extensions.Microsoft.DependencyInjection for mappers

AlexBud369 committed:
fix: remove MediatR (12.4.1) and install (11.1.0) to resolve dependies error

AlexBud369 committed:
feat: add validators for Book commands
- Implements CreateBookCommandValidator to ensure valid book creation data (title, author, tags)
- Implements UpdateBookCommandValidator to validate partial book updates
- Implements DeleteBookCommandValidator to check book and user IDs
- Implements UploadBookCoverCommandValidator to validate cover image URL

AlexBud369 committed:
feat: add validators for Book queries
- Implements GetBookByIdQueryValidator to validate book ID
- Implements GetAllBooksQueryValidator to ensure valid pagination and search parameters

AlexBud369 committed:
feat: add validators for User commands
- Implements BlockUserCommandValidator to ensure valid user and admin IDs
- Implements ChangeRoleCommandValidator to validate role assignment
- Implements DeleteUserCommandValidator to secure user deletion

AlexBud369 committed:
feat: add validators for User queries
- Implements GetUserByIdQueryValidator to validate user ID
- Implements GetAllUsersQueryValidator to ensure valid pagination and search parameters

AlexBud369 committed:
feat: add validators for Auth commands
- Implement LoginCommandValidator with email and password validation
- Implement RegisterCommandValidator with email, password and username validation
- Include validation rules for format, length and required fields

AlexBud369 committed:
feat: add validators for Tags queries
- Implement GetAllTagsQueryValidator with pagination and tag name validation:
- Implement GetTagByIdQueryValidator with required ID check

AlexBud369 committed:
feat: add validation for Tag creating command

budnikme commented on src/Application/Validators/Auth/RegisterCommandValidator.cs:
```diff
@@ -0,0 +1,27 @@
+        RuleFor(x => x.Email)
+            .NotEmpty().WithMessage("Email is required")
+            .EmailAddress().WithMessage("Invalid email format")
+            .MaximumLength(100).WithMessage("Email must not exceed 100 characters");
```
Please move all magic numbers and strings to constants. It is also preferable to use localization for all strings

AlexBud369 committed:
refactor: Update validators with localization and constants

AlexBud369 committed:
feat: add LogoutCommandValidator

AlexBud369 commented:
### Enhanced Validators and Localization
- Updated validators to use `IStringLocalizer<SharedResource>` for localized error messages and `DomainConstants` for numbers/strings
- Refactored `CreateBookCommandValidator` and `UpdateBookCommandValidator` to validate `Tags` as `IEnumerable<Guid>`
- Refactored `GetAllBooksQueryValidator` to use `BookFilterDto`
- Added `LogoutCommandValidator` for `UserId` and `RefreshToken` validation

AlexBud369 committed:
refactor: update UploadBookCoverCommandValidator for IFormFile
- Replaced CoverImageUrl validation with CoverImage (IFormFile) validation
- Added checks for file presence, size, and allowed extensions (.jpg, .jpeg, .png)
- Removed URL-specific validation (NotEmpty, MaximumLength, Matches)

budnikme commented on src/Application/Validators/Books/DeleteBookCommandValidator.cs:
```diff
@@ -0,0 +1,16 @@
+    public DeleteBookCommandValidator(IStringLocalizer<SharedResource> localizer)
+    {
+        RuleFor(x => x.Id)
+            .NotEmpty().WithMessage(localizer["BookIdRequired"]);
```
Please use ``localizer.GetString(SharedResources.BookIdrequired)`` instead

AlexBud369 committed:
fix: enhance Auth validators with proper localization and validation
- Replaced IStringLocalizer with IStringLocalizer for consistency
- Updated localizer['Key'] to localizer.GetString(SharedResources.) for localized error messages
- Replaced NotEmpty with NotEqual(Guid.Empty) for UserId in LogoutCommandValidator
- Fixed typos in RegisterCommandValidator.
- Used DomainConstants.User for EmailMaxLength, UsernameMaxLength, PasswordMinLength, RefreshTokenMaxLength."

AlexBud369 committed:
fix: enhance Books validators with proper localization and validation
- Replaced IStringLocalizer with IStringLocalizer
- Updated localizer['Key'] to localizer.GetString(SharedResources.).
- Replaced NotEmpty with NotEqual(Guid.Empty) for Id, UserId, BookId.
- Updated GetAllBooksQueryValidator to use BookSortBy enum (createdAt, title).
- In UploadBookCoverCommandValidator: renamed Id to BookId, used DomainConstants.BookCover.AllowedImageExtensions, added MaxFileSizeBytes validation, split CoverImageRequired/CoverImageEmpty, replaced InvalidCoverImageFormat with InvalidFileExtension

AlexBud369 committed:
fix: enhance Users validators with proper localization and validation
- Replaced IStringLocalizer with IStringLocalizer
- Updated localizer['Key'] to localizer.GetString(SharedResources.)
- Replaced NotEmpty with NotEqual(Guid.Empty) for Id, UserId, AdminId.
- Fixed GetAllUsersQueryValidator to use DomainConstants.User for DefaultPageNumber and MaxPageSize
- Added DomainConstants.User.Roles.User and .Admin for role validation in ChangeRoleCommandValidator

AlexBud369 committed:
fix: enhance Tags validators with proper localization and validation
- Replaced IStringLocalizer with IStringLocalizer
- Updated localizer['Key'] to localizer.GetString(SharedResources.)
- Replaced NotEmpty with NotEqual(Guid.Empty) for Id in GetTagByIdQueryValidator.
- Fixed GetAllTagsQueryValidator to use DomainConstants.Tag for DefaultPageNumber and MaxPageSize

AlexBud369 committed:
Merge branch 'develop' into feature/application-validators
